readchar/readchar_linux.py

```diff
@@ -1,18 +1,78 @@
 # -*- coding: utf-8 -*-
-# Initially taken from:
-# http://code.activestate.com/recipes/134892/
-# Thanks to Danny Yoo
+# Copyright (c) 2014 - 2017 Miguel Ángel García (@magmax9).
+# Licensed under the MIT license.
+# This code based on code snippet from the python 2 docs:
+# https://docs.python.org/2/faq/library.html
+#     #how-do-i-get-a-single-keypress-at-a-time
+
+
 import sys
-import tty
+import os
 import termios
+import fcntl
+from . import key
+
+
+def get_char(test_stream=None):
+    charbuffer = ''
+    readstream = read_char
+
+    if test_stream:
+        def altreader(*args):
+            return test_stream()
+        readstream = altreader
+
+    while True:
+        if charbuffer in key.ESCAPE_SEQUENCES:
+            char1 = readstream(False)
+        else:
+            char1 = readstream()
+
+        # return if escape sequence is finished (or not an escape sequence).
+        if (charbuffer + char1) not in key.ESCAPE_SEQUENCES:
+            # escape sequence complete or not an escape character..
+            return charbuffer + char1
 
+        # handle cases where the steam finished, but looks incomplete -
+        # such as a plain old 'ESC' (\x1b) that is not followed by other
+        # codes:
+        if (charbuffer + char1) == charbuffer:
+            return charbuffer
 
-def readchar():
+        # add character to sequence and continue loop...
+        charbuffer += char1
+
+
+def read_char(blocking=True):
     fd = sys.stdin.fileno()
-    old_settings = termios.tcgetattr(fd)
+    oldterm = termios.tcgetattr(fd)
+
+    newattr = termios.tcgetattr(fd)
+    newattr[3] = newattr[3] & ~termios.ICANON & ~termios.ECHO
+    termios.tcsetattr(fd, termios.TCSANOW, newattr)
+
+    oldflags = fcntl.fcntl(fd, fcntl.F_GETFL)
+    fcntl.fcntl(fd, fcntl.F_SETFL, oldflags | os.O_NONBLOCK)
+    c = ''
     try:
-        tty.setraw(sys.stdin.fileno())
-        ch = sys.stdin.read(1)
+        while blocking and c == '':
```

Comment:
I believe that the IOError is _not supposed_ to cause an exit.  That is why it is the exception.  We know it will occur, but it is not suppose to stop the loop.